src/main/java/slimeknights/tconstruct/library/recipe/casting/CastingRecipeBuilder.java

```diff
@@ -0,0 +1,175 @@
+import net.minecraft.advancements.criterion.RecipeUnlockedTrigger;
+import net.minecraft.data.IFinishedRecipe;
+import net.minecraft.item.Item;
+import net.minecraft.item.crafting.IRecipeSerializer;
+import net.minecraft.item.crafting.Ingredient;
+import net.minecraft.tags.Tag;
+import net.minecraft.util.IItemProvider;
+import net.minecraft.util.ResourceLocation;
+import net.minecraftforge.fluids.FluidStack;
+import net.minecraftforge.registries.ForgeRegistries;
+import slimeknights.tconstruct.library.recipe.AbstractRecipeBuilder;
+import slimeknights.tconstruct.library.recipe.RecipeUtil;
+import slimeknights.tconstruct.smeltery.TinkerSmeltery;
+
+import javax.annotation.Nullable;
+import java.util.function.Consumer;
+
+public class CastingRecipeBuilder extends AbstractRecipeBuilder<CastingRecipeBuilder> {
+  private final Item result;
+  private boolean castConsumed = false;
+  private boolean switchSlots = false;
+  private FluidStack fluidStack = FluidStack.EMPTY;
+  private Ingredient cast = Ingredient.EMPTY;
+  private int coolingTime = 0;
+  private String group;
+  private final CastingRecipeSerializer<?> recipeSerializer;
+
+  private CastingRecipeBuilder(IItemProvider resultIn, CastingRecipeSerializer<?> serializer) {
+    this.result = resultIn.asItem();
+    this.recipeSerializer = serializer;
+  }
+
+  public static CastingRecipeBuilder castingRecipe(IItemProvider resultIn, CastingRecipeSerializer<?> serializer) {
+    return new CastingRecipeBuilder(resultIn, serializer);
+  }
+
+  public static CastingRecipeBuilder basinRecipe(IItemProvider resultIn) {
+    return castingRecipe(resultIn, TinkerSmeltery.basinRecipeSerializer.get());
+  }
+
+  public static CastingRecipeBuilder tableRecipe(IItemProvider resultIn) {
+    return castingRecipe(resultIn, TinkerSmeltery.tableRecipeSerializer.get());
+  }
+
+  public CastingRecipeBuilder setFluid(FluidStack fluidStack) {
+    this.fluidStack = fluidStack;
+    return this;
+  }
+
+  public CastingRecipeBuilder setGroup(String groupIn) {
+    this.group = groupIn;
+    return this;
+  }
+
+  public CastingRecipeBuilder setCast(Tag<Item> tagIn, boolean consumed) {
+    return this.setCast(Ingredient.fromTag(tagIn), consumed);
+  }
+
+  public CastingRecipeBuilder setCast(IItemProvider itemIn, boolean consumed) {
+    return this.setCast(Ingredient.fromItems(itemIn), consumed);
+  }
+
+  public CastingRecipeBuilder setCast(Ingredient ingredient, boolean consumed) {
+    this.cast = ingredient;
+    this.castConsumed = consumed;
+    return this;
+  }
+
+  /**
+   * Set output of recipe to be put into the input slot.
+   * Mostly used for 'casts'
+   */
+  public CastingRecipeBuilder setSwitchSlots() {
+    this.switchSlots = true;
+    return this;
+  }
+
+  public void build(Consumer<IFinishedRecipe> consumerIn) {
+    this.build(consumerIn, ForgeRegistries.ITEMS.getKey(this.result));
+  }
+
+  public void build(Consumer<IFinishedRecipe> consumerIn, String save) {
```

Comment:
Don't really need this, since we never save recipes to the `minecraft` namespace